src/RedirectMiddleware.php

```diff
@@ -98,7 +99,18 @@ public function checkRedirect(
             );
         }
 
-        return $this($nextRequest, $options);
+        /** @var PromiseInterface|ResponseInterface $response */
+        $options['_previous_uris'][] = $request->getUri();
+        $response = $this($nextRequest, $options);
+
+        // Add headers to be able to track history of redirects.
+        if (!empty($options['allow_redirects']['track_redirects'])) {
+            $response = $response->wait(TRUE);
```

Comment:
I tried really hard to use a promise, but I simply cannot figure it out properly.
The problem is that the most outer promise will need to have the information of all redirects. On the other hand the most outer promise is created within the scope of none already made redirect.

The most inner promise though has the information of all redirects available, but I have absolutely no clue at that point, how to pass down that information. For quite some time I considered the $response from checkRedirect as the previous response, but this is actually not the case. Its, as you know, a response created for manipulation only, something which is designed in a way that the API is independent of the concrete PSR7 implementation.